### ISSUE TITLE
New Scheduler Concept for Weight Decay

### ISSUE THREAD

**Issue by vincentqb** [pytorch/pytorch#22343]
## 🚀 Feature

Weight decay is used very often. A common strategy is simply shrinking the weights at the end of each iteration by a constant multiplicative factor. However, one could expect to use strategies different from this. In that case, we could have weight schedulers that modify the weights using a syntax similar to learning rate schedulers already available. 

## Motivation

There are currently a few issues and pull requests about this, see below.

## Alternatives

* #4429 suggests modifying the optimizer logic to accept a new parameter `weight_decay` specifying the constant multiplicative factor to use.
* #3740, #21250, #22163 introduce variations on Adam and other optimizers with a corresponding built-in weight decay. #3790 is requesting some of these to be supported.
* We could instead have a new "weight_decay_type" option to those optimizers to switch between common strategies.

**Comment by loshchil** [comment]
I would like to add that    

- when the gradient-based and weight decay updates are coupled (the case when L2 regularization is used), then the contribution of weight decay is combined with the gradient update and then multiplied by 'lr',  thus weight decay is scheduled implicitly in the same way as 'lr'.

- when the gradient-based and weight decay updates are decoupled, then there is no multiplication of the weight decay contribution by 'lr' and thus weight decay should be scheduled explicitly (e.g., with the same scheduler as for 'lr', the "coupled" algorithm can be reproduced).

**Comment by PhilJd** [comment]
As a another option, this could be done similarly as we did in tensorflow: Create a factory function that creates a class that executes the optimizer step and afterwards the weight decay step. The advantage is that any optimizer can be extended with decoupled weight decay. It also removes code duplication. 
The new optimizers could then just be named with the suffix `W`.

[Here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/weight_decay_optimizers.py)'s the tf code for reference ;-)

If there's interest, I'd be happy to contribute.

**Comment by vincentqb** [comment]
> As a another option, this could be done similarly as we did in tensorflow: Create a factory function that creates a class that executes the optimizer step and afterwards the weight decay step. The advantage is that any optimizer can be extended with decoupled weight decay. It also removes code duplication.
> The new optimizers could then just be named with the suffix `W`.
> 
> [Here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/opt/python/training/weight_decay_optimizers.py)'s the tf code for reference ;-)
> 
> If there's interest, I'd be happy to contribute.

Thanks for the input: feel free to submit a PR for this :)

**Comment by PhilJd** [comment]
Okay, I've got a first draft ready for AdamW/SGDW, with unscheduled decay (no PR yet).
Regarding schedulers, I think adding a `weight_decay_type` to the optimizer might be a bit unflexible.
How about refactoring the `_LRScheduler(optimizer, last_epoch=-1)` to a `_Scheduler(optimizer, parameter, last_epoch=-1)` class that can schedule arbitrary elements in the `optimizer.param_groups`?
E.g., to schedule learning rate and decay one would do `_Scheduler(optimizer, ["lr", "weight_decay"])`. 
To keep backwards compatibility we could subclass the Schedulers with the `parameter` fixed to `lr`.

What do you think?

**Comment by vincentqb** [comment]
> What do you think?

I see `_Scheduler(optimizer, ["lr", "weight_decay"])` equivalent to having

* `ExponentialLRScheduler`, and
* `ExponentialWScheduler`

Could we have a situation were we have the following schedulers?

* `ExponentialLRScheduler`, and
* `StepWScheduler`

In this case, I would expect to specify `_Scheduler(optimizer, "lr") and _Scheduler(optimizer, "weight_decay")`.

By the way, there's a discussion going on about chaining schedulers in #21800 that will impact how a general scheduler would work. Let's see what comes out of it, so we can keep this work in sync. In the meanwhile, let me write down the current thinking. We want:

* An abstract scheduler class that can act on any one of the parameter (learning rate, weight, etc.), as you mention: `_Scheduler(optimizer, parameter, last_epoch=-1)`.
* All the current learning rate scheduler would simply become children of these classes, targeting the learning rate parameter. And we can create child that act on weights, or whatever else.
* The current behavior in the optimizer should naturally be maintained if a user only uses a learning rate scheduler. (What happens if a user specify only a weight/generic scheduler?)

Thoughts?

**Comment by veritas9872** [comment]
@vincentqb Hello. I think that this is a great idea? Has there been any progress?
